### PR TITLE
TELCODOCS-329: MachineHealthCheck: Implement metal3 remediation using new external API

### PR DESCRIPTION
**Pull request by ogradyp**
https://issues.redhat.com/browse/TELCODOCS-329 MachineHealthCheck: Implement metal3 remediation using new external API

Applies to OpenShift version : 4.13 +

Preview:
[MachineHealthChecks on bare metal](https://57986--docspreview.netlify.app/openshift-enterprise/latest/machine_management/deploying-machine-health-checks.html#machine-health-checks-bare-metal_deploying-machine-health-checks)
[Understanding the annotation-based remediation process](https://57986--docspreview.netlify.app/openshift-enterprise/latest/machine_management/deploying-machine-health-checks.html#mgmt-understanding-remediation-process_deploying-machine-health-checks)
[Understanding the metal3-based remediation process](https://57986--docspreview.netlify.app/openshift-enterprise/latest/machine_management/deploying-machine-health-checks.html#mgmt-understanding-metal3-remediation-process_deploying-machine-health-checks)
[Creating a MachineHealthCheck resource for bare metal](https://57986--docspreview.netlify.app/openshift-enterprise/latest/machine_management/deploying-machine-health-checks.html#mgmt-creating-mhc-baremetal_deploying-machine-health-checks)

Dev review completed by @slintes 
QE review completed by @anna-savina 
Peer review completed by @adellape 

Thank you.

### CODE REVIEW

File: modules/mgmt-power-remediation-baremetal-about.adoc [hunk 1]
```diff
@@ -105,6 +125,41 @@ spec:
 The `matchLabels` are examples only; you must map your machine groups based on your specific needs.
 ====
 
+.Sample `MachineHealthCheck` resource for bare metal, metal3-based remediation
```
Comment (anna-savina) [inline modules/mgmt-power-remediation-baremetal-about.adoc:128]: Maybe also to add example of `metal3-remediation-template`?

```
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: Metal3RemediationTemplate
metadata:
  name: metal3-remediation
  namespace: openshift-machine-api
spec:
  template:
    spec:
      strategy:
        retryLimit: 1
        timeout: 5m0s
        type: Reboot
```

Comment (slintes) [reply modules/mgmt-power-remediation-baremetal-about.adoc:128]: yes please!

File: modules/mgmt-power-remediation-baremetal-about.adoc [hunk 2]
```diff
@@ -23,15 +23,18 @@ Power-based remediation provides the following capabilities:
 Machine deletion on bare metal cluster triggers reprovisioning of a bare metal host.
 Usually bare metal reprovisioning is a lengthy process, during which the cluster
 is missing compute resources and applications might be interrupted.
-To change the default remediation process from machine deletion to host power-cycle,
-annotate the `MachineHealthCheck` resource with the
+
+Thee are two ways to change the default remediation process from machine deletion to host power-cycle:
```
Comment (slintes) [inline modules/mgmt-power-remediation-baremetal-about.adoc:27]: typo: There

File: modules/mgmt-power-remediation-baremetal-about.adoc (hunk 1 above)
Comment (ogradyp) [reply modules/mgmt-power-remediation-baremetal-about.adoc:128]: Thanks for your feedback @anna-savina. I have applied the fix.

Comment (anna-savina) [reply modules/mgmt-power-remediation-baremetal-about.adoc:128]: @ogradyp Sorry for misleading. I meant to add example of metal3-remediation-template to existing MHC CR::

.Sample `MachineHealthCheck` resource for bare metal, metal3-based remediation

```
apiVersion: machine.openshift.io/v1beta1
kind: MachineHealthCheck
metadata:
  name: example
  namespace: openshift-machine-api
spec:
  selector:
    matchLabels:
      machine.openshift.io/cluster-api-machine-role: <role>
      machine.openshift.io/cluster-api-machine-type: <role>
      machine.openshift.io/cluster-api-machineset: <cluster_name>-<label>-<zone>
  selector:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: Metal3RemediationTemplate
    name: metal3-remediation-template
    namespace: openshift-machine-api
  unhealthyConditions:
  - type:    "Ready"
    timeout: "300s"
```

.Sample `Metal3RemediationTemplate` resource for bare metal, metal3-based remediation

```
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: Metal3RemediationTemplate
metadata:
  name: metal3-remediation-template
  namespace: openshift-machine-api
spec:
  template:
    spec:
      strategy:
        retryLimit: 1
        timeout: 5m0s
        type: Reboot
```

Comment (ogradyp) [reply modules/mgmt-power-remediation-baremetal-about.adoc:128]: Not at all @anna-savina . Thanks for your feedback. I have added the example of metal3-remediation-template to existing MHC CR.

File: modules/mgmt-power-remediation-baremetal-about.adoc [hunk 3]
```diff
@@ -105,6 +125,52 @@ spec:
...
+kind: Metal3RemediationTemplate
+metadata:
+  name: metal3-remediation-template
+  namespace: openshift-machine-api
+spec:
+  template:
+    spec:
+      strategy:
+        type: "Reboot"
+        retryLimit: "1"
```
[the remark this answers is not in the document]
Comment (ogradyp) [reply modules/mgmt-power-remediation-baremetal-about.adoc]: Thanks for your feedback @slintes. I have applied the fix.

**Comment by slintes** [comment]
lgtm

**Comment by anna-savina** [comment]
lgtm

File: modules/mgmt-power-remediation-baremetal-about.adoc [hunk 4]
```diff
@@ -47,6 +50,23 @@ The remediation process operates as follows:
 If the power operations did not complete, the bare metal machine controller triggers the reprovisioning of the unhealthy node unless this is a control plane node or a node that was provisioned externally.
 ====
 
+[id="mgmt-understanding-metal3-remediation-process_{context}"]
+== Understanding the metal3-based remediation process
+
+The remediation process operates as follows:
+
+. The MachineHealthCheck (MHC) controller detects that a node is unhealthy.
+. The MHC creates a metal3 remediation custom resource for the metal3 remediation controller which requests to power-off the unhealthy node.
```
Comment (adellape) [inline modules/mgmt-power-remediation-baremetal-about.adoc:59]:
```suggestion
. The MHC creates a metal3 remediation custom resource for the metal3 remediation controller, which requests to power-off the unhealthy node.
```

File: modules/mgmt-power-remediation-baremetal-about.adoc [hunk 5]
```diff
@@ -105,6 +125,52 @@ spec:
 The `matchLabels` are examples only; you must map your machine groups based on your specific needs.
 ====
 
+.Sample `MachineHealthCheck` resource for bare metal, metal3-based remediation
+[source,yaml]
+----
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  name: example <1>
```
Comment (adellape) [inline modules/mgmt-power-remediation-baremetal-about.adoc:134]: Seems to be missing a matching `<1>` callout description below?

File: modules/mgmt-power-remediation-baremetal-about.adoc (hunk 4 above)
Comment (ogradyp) [reply modules/mgmt-power-remediation-baremetal-about.adoc]: Thanks for your feedback @adellape . I have applied the fix.

File: modules/mgmt-power-remediation-baremetal-about.adoc (hunk 5 above)
Comment (ogradyp) [reply modules/mgmt-power-remediation-baremetal-about.adoc]: Thanks for your feedback @adellape . I have removed the <1>.

**Comment by adellape** [comment]
@ogradyp Changes look good, but we need to squash the 8 commits down to 1 before merge. Dropping merge labels for now; please re-queue when ready. Thank you!

**Comment by ogradyp** [comment]
> /label merge-review-needed

> @ogradyp Changes look good, but we need to squash the 8 commits down to 1 before merge. Dropping merge labels for now; please re-queue when ready. Thank you!

Thanks for your feedback @adellape . I have squashed the 8 commits down to 1.

**Review by kelbrown20 — COMMENTED** [review]
Updates LGTM! 
Merging

**Comment by ogradyp** [comment]
Thank you for merging @kelbrown20

**Pull request merged**